### ISSUE TITLE
NOT_NULL not validating properly?

### ISSUE THREAD

**Issue by autoferrit** [3YOURMIND/django-migration-linter#104]
I have a model as such, where I am adding ONLY the `description` field. Which should allow null and blank values:
```python
class BlogPost(models.Model):
    """
    This model is used just as an example.

    With it we show how one can:
    - Use fixtures and factoris
    - Use migrations testing

    """

    title = models.CharField(max_length=_POST_TITLE_MAX_LENGTH)
    body = models.TextField()
    # only this line is new
    description = models.TextField(null=True, blank=True)

    created_at = models.DateTimeField(auto_now_add=True)
    modifiet_at = models.DateTimeField(auto_now=True)

    def __str__(self) -> str:
        """All django models should have this method."""
        return textwrap.wrap(self.title, _POST_TITLE_MAX_LENGTH // 4)[0]
```
and running `makemigrations` creates this migration:
```python
from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('main', '0001_initial'),
    ]

    operations = [
        migrations.AddField(
            model_name='blogpost',
            name='description',
            field=models.TextField(blank=True, null=True),
        ),
    ]
```
and the migration seems correct. I am adding a new column which is backwards compatible by allowing NULL and blank values. But when running `./manage.py lintmigrations` I get this error:
```bash
➜ ./manage.py lintmigrations     
(axes, 0001_initial)... OK (cached)
(axes, 0002_auto_20151217_2044)... ERR (cached)
	NOT NULL constraint on columns
(axes, 0003_auto_20160322_0929)... ERR (cached)
	NOT NULL constraint on columns
(axes, 0004_auto_20181024_1538)... ERR (cached)
	NOT NULL constraint on columns
(axes, 0005_remove_accessattempt_trusted)... ERR (cached)
	NOT NULL constraint on columns
(axes, 0006_remove_accesslog_trusted)... ERR (cached)
	NOT NULL constraint on columns
(db, 0001_initial)... OK (cached)
(main, 0001_initial)... OK (cached)
(main, 0002_blogpost_description)... ERR
	NOT NULL constraint on columns
*** Summary:
Valid migrations: 3/9 - erroneous migrations: 6/9 - ignored migrations: 0/9
```
Where it is giving me a NOT NULL constraint. Which  should not be true as I want to allow it to be nullable. the new migration is `0002_blogpost_description` So I cannot figure out why I am getting this error.

**Comment by autoferrit** [comment]
Adding `default=None` to the field also has no effect

**Comment by David-Wobrock** [comment]
Hi @autoferrit , thanks for submitting an issue.
Weird :thinking:  could you please provide some details about:
* the Python version you are using
* the Django version
* the database you are targeting
* and ideally, the output of `./manage.py sqlmigrate main 0002_blogpost_description`

So that I'm able to reproduce

**Comment by David-Wobrock** [comment]
Hey @autoferrit 

To re-open the discussion, it seems that adding a nullable `TextField` gives the following SQL depending on the DB (app: `app_add_not_null_column_a`, the field is called `f`):
* postgresql
```
ALTER TABLE "app_add_not_null_column_a" ADD COLUMN "f" text NULL;
```
* mysql
```
ALTER TABLE `app_add_not_null_column_a` ADD COLUMN `f` longtext NULL;
```
* sqlite
```
CREATE TABLE "new__app_add_not_null_column_a" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "f" text NULL, "null_field" integer NULL, "new_not_null_field" integer NOT NULL);
INSERT INTO "new__app_add_not_null_column_a" ("id", "null_field", "new_not_null_field", "f") SELECT "id", "null_field", "new_not_null_field", NULL FROM "app_add_not_null_column_a";
DROP TABLE "app_add_not_null_column_a";
ALTER TABLE "new__app_add_not_null_column_a" RENAME TO "app_add_not_null_column_a";
COMMIT;
```

So the issue is probably linked to the fact that you're using `sqlite`, for which Django creates a new table, copies data and renames it in order to add a new column.

I agree that the detection here should probably be more fine-grained, to detect that the added column in the new table won't cause trouble.

**Comment by maximekl** [comment]
Hi,

Django documentation said to avoid using `null=True` on `CharField` and `TextField` : https://docs.djangoproject.com/en/3.0/ref/models/fields/#null

Use : `default=""` instead of `default=None`

**Comment by autoferrit** [comment]
Sorry I was doing this for a new project, and I eventually just didnt use this lib. But, I wasn't using sqlite, i was using:
* Postgres: 12
* Python 3.8
* Django 3

I will have to try to get to a point where I can run that command again but i haven't touched the project much lately.

**Comment by David-Wobrock** [comment]
No activity, not using the tool - closing this issue.
Don't hesitate to re-open if we have new information

**Comment by kirsinger** [comment]
I'm experiencing a very similar looking error while trying to add a not-null `TextField` to an existing model with a default value set. The model looks like this:
```
class PurchaseOrder(models.Model):
    # ... existing fields ...
    narration = TextField(default='')
```
And the warning raised by the linter looks like this:

![image](https://user-images.githubusercontent.com/7554304/133964174-3b2d642a-a0a4-4cb8-8116-471fb0b8fb95.png)

From the SQL in that error message, it looks like a default value is being set for the new field, so I'm not quite sure what the problem is here. I'm using Django 3.2 on Python 3.8 backed by Postgres 11.

**Comment by David-Wobrock** [comment]
Hi @kirsinger 

> From the SQL in that error message, it looks like a default value is being set for the new field, so I'm not quite sure what the problem is here. I'm using Django 3.2 on Python 3.8 backed by Postgres 11.

The linter working as expected, and is detecting exactly one of the pitfalls in Django.
From your screenshot, adding the `narration = TextField(default='')` field is generating two SQL statements:
First, adding the new column, not nullable with your default value **to fill existing rows**:
```
ALTER TABLE payments_purchaseorder ADD COLUMN "narration" text DEFAULT \'\' NOT NULL;
```
**Secondly, dropping the default value**
```
ALTER TABLE payments_purchaseorder ALTER COLUMN "narration" DROP DEFAULT;
```

You find yourself in the position where you added a NOT NULL column without a default value.
This is not backwards compatible, because an old code base, that isn't aware of the `narration` field wouldn't specify it when creating a `PurchaseOrder`, but the database doesn't accept the not null (and the default has been dropped).

You can read more about this here https://medium.com/botify-labs/django-and-its-default-values-c21a13cff9f
